File: app/domain/repositories/usage_repository.py
"""
Usage records repository — handles all CRUD and statistics queries for `usage_records` table.
"""
from datetime import datetime
from app.infrastructure.database.connection import ConnectionManager


class UsageRepository:
    """Data access for usage_records table."""

    def __init__(self, conn_mgr: ConnectionManager):
        self._conn_mgr = conn_mgr

    def add_usage_record(self, user_name: str, start_time: datetime, end_time: datetime) -> int:
        """Add a usage record. Returns the new record ID."""
        duration = int((end_time - start_time).total_seconds())
        with self._conn_mgr.connection() as conn:
            cursor = conn.cursor()
            cursor.execute(
                "INSERT INTO usage_records (user_name, start_time, end_time, duration_seconds) VALUES (?, ?, ?, ?)",
                (user_name, start_time.strftime("%Y-%m-%d %H:%M:%S"),
                 end_time.strftime("%Y-%m-%d %H:%M:%S"), duration)
            )
            conn.commit()
            return cursor.lastrowid

    def get_usage_records(self, filter_mode: str = "all") -> list:
        """Get usage records filtered by mode ('今日', '本月', 'all')."""
        with self._conn_mgr.connection() as conn:
            cursor = conn.cursor()
            if filter_mode == "今日":
                cursor.execute(
                    "SELECT * FROM usage_records WHERE date(start_time) = date('now', 'localtime') ORDER BY start_time DESC"
                )
            elif filter_mode == "本月":
                cursor.execute(
                    "SELECT * FROM usage_records WHERE strftime('%Y-%m', start_time) = strftime('%Y-%m', 'now', 'localtime') ORDER BY start_time DESC"
                )
            else:
                cursor.execute("SELECT * FROM usage_records ORDER BY start_time DESC")
            return [dict(row) for row in cursor.fetchall()]

    def get_today_records(self) -> list:
        """Get today's usage records."""
        return self.get_usage_records("今日")

    def get_user_stats(self, filter_mode: str = "all") -> list:
        """Get aggregated user statistics."""
        with self._conn_mgr.connection() as conn:
            cursor = conn.cursor()
            if filter_mode == "今日":
                date_filter = "AND date(start_time) = date('now', 'localtime')"
            elif filter_mode == "本月":
                date_filter = "AND strftime('%Y-%m', start_time) = strftime('%Y-%m', 'now', 'localtime')"
            else:
                date_filter = ""
            cursor.execute(f'''
                SELECT user_name,
                       SUM(duration_seconds) as total_seconds,
                       MAX(end_time) as last_seen
                FROM usage_records
                WHERE 1=1 {date_filter}
                GROUP BY user_name
                ORDER BY total_seconds DESC
            ''')
            return [dict(row) for row in cursor.fetchall()]

    def delete_usage_record(self, record_id: int):
        """Delete a single usage record."""
        with self._conn_mgr.connection() as conn:
            cursor = conn.cursor()
            cursor.execute("DELETE FROM usage_records WHERE id = ?", (record_id,))
            conn.commit()

    def delete_usage_records(self, record_ids: list):
        """Delete multiple usage records."""
        if not record_ids:
            return
        with self._conn_mgr.connection() as conn:
            cursor = conn.cursor()
            placeholders = ",".join("?" * len(record_ids))
            cursor.execute(f"DELETE FROM usage_records WHERE id IN ({placeholders})", record_ids)
            conn.commit()

    def get_hourly_stats(self) -> list:
        """Get hourly usage statistics for heatmap."""
        with self._conn_mgr.connection() as conn:
            cursor = conn.cursor()
            cursor.execute('''
                SELECT date(start_time) as date,
                       CAST(strftime('%H', start_time) as INTEGER) as hour,
                       SUM(duration_seconds) / 3600.0 as hours
                FROM usage_records
                GROUP BY date(start_time), strftime('%H', start_time)
            ''')
            return [dict(row) for row in cursor.fetchall()]

    def get_daily_stats(self, days: int = 30) -> list:
        """Get daily usage statistics for trend chart."""
        with self._conn_mgr.connection() as conn:
            cursor = conn.cursor()
            cursor.execute(f'''
                SELECT date(start_time) as date,
                       SUM(duration_seconds) / 3600.0 as hours,
                       COUNT(DISTINCT user_name) as users
                FROM usage_records
                WHERE date(start_time) >= date('now', 'localtime', '-{days} days')
                GROUP BY date(start_time)
                ORDER BY date
            ''')
            return [dict(row) for row in cursor.fetchall()]

    def get_user_distribution(self, filter_mode: str = "all") -> list:
        """Get user distribution for pie chart."""
        return self.get_user_stats(filter_mode)

File: src/Monitor.App/Repositories/UsageRepository.cs
using Dapper;
using Monitor.App.Infrastructure.Database;
using Monitor.App.Models;

namespace Monitor.App.Repositories;

public class UsageRepository : IUsageRepository
{
    private readonly DbConnectionFactory _factory;

    public UsageRepository(DbConnectionFactory factory) => _factory = factory;

    public int AddUsageRecord(string userName, string startTime, string endTime)
    {
        using var conn = _factory.CreateConnection();
        var start = DateTime.Parse(startTime);
        var end = DateTime.Parse(endTime);
        var duration = (int)(end - start).TotalSeconds;

        var sql = "INSERT INTO usage_records (user_name, start_time, end_time, duration_seconds) VALUES (@Name, @Start, @End, @Duration); SELECT last_insert_rowid();";
        return conn.ExecuteScalar<int>(sql, new { Name = userName, Start = startTime, End = endTime, Duration = duration });
    }

    public List<UseHistoryItem> GetUsageRecords(string filterMode = "all")
    {
        using var conn = _factory.CreateConnection();
        var sql = "SELECT id, user_name AS UserName, start_time AS StartTime, end_time AS EndTime, duration_seconds AS DurationSeconds, created_at AS CreatedAt FROM usage_records";
        if (filterMode == "today")
            sql += " WHERE date(start_time) = date('now', 'localtime')";
        else if (filterMode == "this month")
            sql += " WHERE strftime('%Y-%m', start_time) = strftime('%Y-%m', 'now', 'localtime')";
        sql += " ORDER BY start_time DESC";
        return conn.Query<UseHistoryItem>(sql).AsList();
    }

    public List<UseHistoryItem> GetTodayRecords() => GetUsageRecords("today");

    public List<UserStat> GetUserStats(string filterMode = "all")
    {
        using var conn = _factory.CreateConnection();
        var sql = "SELECT user_name AS UserName, SUM(duration_seconds) AS TotalSeconds, MAX(end_time) AS LastSeen FROM usage_records";
        if (filterMode == "today")
            sql += " WHERE date(start_time) = date('now', 'localtime')";
        else if (filterMode == "this month")
            sql += " WHERE strftime('%Y-%m', start_time) = strftime('%Y-%m', 'now', 'localtime')";
        sql += " GROUP BY user_name ORDER BY TotalSeconds DESC";
        return conn.Query<UserStat>(sql).AsList();
    }

    public void DeleteUsageRecord(int id)
    {
        using var conn = _factory.CreateConnection();
        conn.Execute("DELETE FROM usage_records WHERE id = @Id", new { Id = id });
    }

    public void DeleteUsageRecords(List<int> ids)
    {
        if (ids.Count == 0) return;
        using var conn = _factory.CreateConnection();
        conn.Execute($"DELETE FROM usage_records WHERE id IN ({string.Join(",", ids)})");
    }

    public List<HourlyStat> GetHourlyStats()
    {
        using var conn = _factory.CreateConnection();
        var sql = @"
            SELECT date(start_time) AS Date,
                   CAST(strftime('%H', start_time) AS INTEGER) AS Hour,
                   ROUND(SUM(duration_seconds) / 3600.0, 2) AS Hours
            FROM usage_records
            GROUP BY date(start_time), strftime('%H', start_time)
            ORDER BY date(start_time), Hour";
        return conn.Query<HourlyStat>(sql).AsList();
    }

    public List<DailyStat> GetDailyStats(int days = 30)
    {
        using var conn = _factory.CreateConnection();
        var sql = @"
            SELECT date(start_time) AS Date,
                   ROUND(SUM(duration_seconds) / 3600.0, 2) AS Hours,
                   COUNT(DISTINCT user_name) AS Users
            FROM usage_records
            WHERE start_time >= date('now', 'localtime', @Days)
            GROUP BY date(start_time)
            ORDER BY date(start_time)";
        return conn.Query<DailyStat>(sql, new { Days = $"-{days} days" }).AsList();
    }

    public List<UserStat> GetUserDistribution(string filterMode = "all") => GetUserStats(filterMode);
}
